sys/dev/usb/usb_generic.c

```diff
@@ -1192,54 +1203,175 @@ ugen_fs_copy_in(struct usb_fifo *f, uint8_t ep_index)
 			}
 
 			/* move data */
-			error = copyin(uaddr, kaddr, length);
+			error = copyin(fs_ep.ppBuffer[n], kaddr, length);
 			if (error) {
 				break;
 			}
 		}
 		offset += length;
 	}
+err:
+	free(fs_ep.ppBuffer, M_USB);
 	return (error);
 
 complete:
+	free(fs_ep.ppBuffer, M_USB);
 	mtx_lock(f->priv_mtx);
 	ugen_fs_set_complete(f, ep_index);
 	mtx_unlock(f->priv_mtx);
 	return (0);
 }
 
+static struct usb_fs_endpoint *
+ugen_fs_ep_uptr(struct usb_fifo *f, uint8_t ep_index)
+{
+	return ((struct usb_fs_endpoint *)
+	    ((char *)f->fs_ep_ptr + (ep_index * f->fs_ep_sz)));
+}
+
+static int
+ugen_fs_copyin(struct usb_fifo *f, uint8_t ep_index,
+    struct usb_fs_endpoint* fs_ep, bool copy_arrays)
+{
+	void *uptr;
+#ifdef COMPAT_FREEBSD32
+	struct usb_fs_endpoint32 fs_ep32;
+	uint32_t *uptr32;
+#endif
+	int error;
+
+	switch(f->fs_ep_sz) {
+	case sizeof(struct usb_fs_endpoint):
```

Comment:
Is there a reason you chose this over if/else if/else? Not aware of this being a common pattern in FreeBSD so far?

Reply:
There will be three values in CheriBSD so a case is more natural and this will produce less churn.